src/main/java/org/jetbrains/bsp/bazel/server/BepServer.java

```diff
@@ -88,156 +86,146 @@ public void publishLifecycleEvent(
+      throws IOException {
+    if (event.hasAction()) {
+      processActionEvent(event.getAction());
+    }
+  }
+
+  private void processAbortedEventIfPresent(BuildEventStreamProtos.BuildEvent event) {
+    if (event.hasAborted()) {
+      processAbortedEvent(event.getAborted());
+    }
+  }
+
+  private void processProgressEventIfPresent(BuildEventStreamProtos.BuildEvent event) {
+    if (event.hasProgress()) {
+      processProgressEvent(event.getProgress());
+    }
+  }
+
+  private void processedStartedEvent(BuildEventStreamProtos.BuildStarted buildStarted) {
+    TaskId taskId = new TaskId(buildStarted.getUuid());
+    TaskStartParams startParams = new TaskStartParams(taskId);
+    startParams.setEventTime(buildStarted.getStartTimeMillis());
+
+    bspClient.onBuildTaskStart(startParams);
+    taskParkingLot.add(taskId);
+  }
+
+  private void processFinishedEvent(BuildEventStreamProtos.BuildFinished buildFinished) {
+    if (taskParkingLot.empty()) {
+      LOGGER.info("No start event id was found.");
+      return;
+    }
+
+    if (taskParkingLot.size() > 1) {
+      LOGGER.info("More than 1 start event was found");
+      return;
+    }
+
+    StatusCode exitCode = BepServer.convertExitCode(buildFinished.getExitCode().getCode());
+    TaskFinishParams finishParams = new TaskFinishParams(taskParkingLot.pop(), exitCode);
+    finishParams.setEventTime(buildFinished.getFinishTimeMillis());
+
+    bspClient.onBuildTaskFinish(finishParams);
+  }
+
+  private void processProgressEvent(BuildEventStreamProtos.Progress progress) {
+    Map<String, List<Diagnostic>> fileDiagnostics = parseStdErrDiagnostics(progress);
+
+    fileDiagnostics.entrySet().stream()
+        .map(this::createParamsFromEntry)
+        .forEach(bspClient::onBuildPublishDiagnostics);
 
-    fileDiagnostics.forEach(
-        (fileLocation, diagnostics) ->
-            bspClient.onBuildPublishDiagnostics(
-                new PublishDiagnosticsParams(
-                    new TextDocumentIdentifier(Uri.fromAbsolutePath(fileLocation).toString()),
-                    new BuildTargetIdentifier(""),
-                    diagnostics,
-                    false)));
+    buildClientLogger.logMessage(progress.getStderr().trim());
   }
 
-  private void processCompletionEvent(BuildEventStreamProtos.BuildEvent event) throws IOException {
-    List<BuildEventStreamProtos.OutputGroup> outputGroups =
-        event.getCompleted().getOutputGroupList();
+  private PublishDiagnosticsParams createParamsFromEntry(
+      Map.Entry<String, List<Diagnostic>> entry) {
+    String fileLocation = entry.getKey();
+    List<Diagnostic> diagnostics = entry.getValue();
+
+    return new PublishDiagnosticsParams(
+        new TextDocumentIdentifier(Uri.fromAbsolutePath(fileLocation).toString()),
+        new BuildTargetIdentifier(""),
+        diagnostics,
+        false);
+  }
+
+  private void processCompletedEvent(BuildEventStreamProtos.TargetComplete targetComplete)
+      throws IOException {
```

Comment:
Again I think this IO exception should be wrapped inside private methods

Reply:
I'll try to wrap all the IOExceptions while refactoring those methods more thoroughly later.